xarray/tutorial.py

```diff
@@ -6,23 +6,78 @@
 
 """
 import hashlib
-import os as _os
-from urllib.request import urlretrieve
+import pathlib
+import shutil
+import tempfile
 
 import numpy as np
+import requests
 
 from .backends.api import open_dataset as _open_dataset
+from .backends.rasterio_ import open_rasterio as _open_rasterio
 from .core.dataarray import DataArray
 from .core.dataset import Dataset
 
-_default_cache_dir = _os.sep.join(("~", ".xarray_tutorial_data"))
+_cache_name = "xarray_tutorial_data"
+_cache_dir = pathlib.Path.home() / ".cache"
+# TODO: I/O on import. Might not be a good idea.
+if _cache_dir.exists():
+    _default_cache_dir = _cache_dir / _cache_name
+else:
+    _default_cache_dir = pathlib.Path.home() / f".{_cache_name}"
 
 
-def file_md5_checksum(fname):
-    hash_md5 = hashlib.md5()
-    with open(fname, "rb") as f:
-        hash_md5.update(f.read())
-    return hash_md5.hexdigest()
+def check_md5sum(content, checksum):
+    md5 = hashlib.md5()
+    md5.update(content)
+    md5sum = md5.hexdigest()
+
+    return md5sum == checksum
+
+
+def download_to(url, path):
+    with requests.get(url, stream=True) as r, path.open("wb") as f:
+        if r.status_code != 200:
+            raise OSError(f"download failed: {r.reason}")
+
+        r.raw.decode_content = True
+        shutil.copyfileobj(r.raw, f)
+
+
+def open_rasterio(
+    name,
+    cache=True,
+    cache_dir=_default_cache_dir,
+    github_url="https://github.com/mapbox/rasterio",
+    branch="master",
+    **kws,
+):
+    if not cache_dir.is_dir():
+        cache_dir.mkdir()
+
+    default_extension = ".tif"
+
+    if cache:
+        path = cache_dir / name
+        # need to always do that, otherwise the context manager might fail
+        cache_dir = pathlib.Path(cache_dir)
+    else:
+        cache_dir = tempfile.TemporaryDirectory()
+        path = pathlib.Path(cache_dir.name) / name
+
+    if not path.suffix:
+        path = path.with_suffix(default_extension)
+    elif path.suffix == ".byte":
+        path = path.with_name(name + default_extension)
+
+    if cache and path.is_file():
+        return _open_rasterio(path, **kws)
+
+    url = f"{github_url}/raw/{branch}/tests/data/{path.name}"
+    # make sure the directory is deleted afterwards
+    with cache_dir:
+        download_to(url, path)
```

Comment:
The reason why the MD5 hash is checked below is to guard against incomplete / failed downloads. This is actually a pretty common scenario, e.g., if a user hits Ctrl+C because downloading is taking too long (e.g., if they have a poor internet connection). So I think this would be important to include for rasterio as well.

That said, I'm not sure we really need to use MD5 verification, which is rather annoying to setup and requires downloading extra files. Instead, I would suggest ensuring that `download_to` always deletes the downloaded file if an exception is raised, e.g., based on https://stackoverflow.com/a/27045091/809705:
```python
def download_to(url, path):
    try:
        ...  # current contents of `download_to`
    except Exception:
        # switch to path.unlink(missing_ok=True) for Python 3.8+
        with contextlib.suppress(FileNotFoundError):
            path.unlink()
        raise
```

This would probably work in about 90% of cases.

There may be a few others where a computer is shut down suddenly in the middle to writing a file. To handle this, it's probably a good idea to try to do an _atomic write_:
https://stackoverflow.com/a/2333979/809705

Reply:
I implemented the "atomic" write with a deterministic temporary name, which (as long as we properly detect something like disconnected sockets / broken pipes) should guard against incomplete writes and also avoid stale files. The only disadvantage I know of is that concurrent calls will interfere with each other.

Reply:
My main concern here would that if two invocations of `tutorial.open_dataset` start at the same time (e.g., in different processes). Then it's possible that both could partially write to the same temporary file, which could get moved into place with an inconsistent state.

I think it's probably a better idea to make the temporary file with a random name (like the [atomicwrites](https://github.com/untitaker/python-atomicwrites) package). In the worst case you end up with an extra file, but that's better than an bad state that requires manual intervention to resolve.

Another option is to vendor a copy of [atomicwrites](https://github.com/untitaker/python-atomicwrites) (e.g., as well as appdirs). It's a single file project so this would be relatively straightforward.

Reply:
sure. I changed it to use `NamedTemporaryFile` to generate a unique file within the same directory and rename once it's done downloading. If the downloading succeeded, the rename will always work (not sure if that's the case on windows, though)